src/inventory/slot_mapping.rs

```diff
@@ -0,0 +1,196 @@
+use crate::inventory::{Item, Slot};
+use crate::render::hud::Hud;
+use crate::render::inventory::InventoryWindow;
+use crate::render::Renderer;
+use crate::ui::{Container, VAttach};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use parking_lot::RwLock;
+
+/// A mapping of each slot and it's position on the screen. This makes the job
+/// of rendering and locating slots easier for inventories. This class only
+/// expects positions in pixels relative to the top-left corner of the window.
+pub struct SlotMapping {
+    /// A mapping of slot_id to (position, slot). The position is the distance
+    /// between the window's top-left corner, and the slot's top-left corner in
+    /// pixels. This doesn't include the border frame around each slot, just
+    /// the 16x16 content.
+    slots: HashMap<u16, ((i32, i32), Slot)>,
+    /// The size of the window in pixels.
+    size: (i32, i32),
+    dirty: bool,
+    /// A child slot mapping to be used as a fallback if a slot wasn't found
+    /// in the parent mapping. This is used to handle the base inventory.
+    child: Option<Arc<RwLock<SlotMapping>>>,
+    /// The offset between the top-left corner of the current mapping window,
+    /// and the corner of the sub-section that the child mapping manages, in
+    /// pixels.
+    child_offset: Option<(i32, i32)>,
```

Comment:
This could probably become a custom struct containing an AtomicU64 which contains the 2 i32s and which has a getter and setter to which 2 i32s get passed/from which 2 i32s get returned

Reply:
only interior mutability comment